# AI VIBECODING MISSION — NEXUS Engineering Solutions
**Case Study Challenge | Python | 5 Challenges | 115 pts**

| Field | Detail |
|-------|--------|
| **Team Name** | NEXUS Squad |
| **Date** | 2026-04-28 |
| **Student 1** | Aqsa Shah |
| **Student 2** | Hina |

---
> **Scenario:** NEXUS Engineering Solutions is a startup that automates factories across Sindh. Their systems are broken, biased, and incomplete. Our team has been hired as AI-powered engineers. We used vibecoding (clear prompts + verification) to fix, extend, and build real tools.

---
## CHALLENGE 1 — Bug Hunt (20 pts)
### Faulty Sensor Monitor — Karachi Factory Floor

**Scenario:** The NEXUS sensor monitor crashes every morning and gives wrong safety alerts.

### Task 1.1 — Diagnose (5 pts)

**BUG 1** — Line: `average = sum(readings) / len(readings)`  
Jab `readings` list empty ho (`[]`), to `len([]) = 0` hota hai. Python zero se divide nahi kar sakta — isliye `ZeroDivisionError` crash ho jata hai.

**BUG 2** — Line: `print(analyse_sensors(data_empty))`  
`data_empty = []` pass ki gayi lekin function mein koi empty-list check nahi tha — isliye BUG 1 automatically trigger hota tha aur poora program band ho jata tha.

### Task 1.2 — Prompt & Fix (10 pts)

**Our AI Prompt:**
> *"I have a Python function `analyse_sensors(readings)` that calculates the average of a list and returns a status dict. It has two bugs: Bug 1 — it crashes with ZeroDivisionError when the list is empty because it does `sum(readings)/len(readings)` with no empty check. Bug 2 — calling it with `data_empty=[]` triggers that crash. Please fix the function so it handles an empty list gracefully and returns `{'average': 0, 'status': 'NO DATA'}` instead of crashing. Keep the rest of the logic the same."*

**AI's Fixed Code (below):**

In [1]:
# CHALLENGE 1 — Fixed Sensor Monitor

def analyse_sensors(readings):
    if not readings:                          # FIX: empty list ko handle karo
        return {'average': 0, 'status': 'NO DATA'}
    average = sum(readings) / len(readings)
    if average > 85:
        status = 'CRITICAL'
    elif average > 60:
        status = 'WARNING'
    else:
        status = 'OK'
    return {'average': round(average, 2), 'status': status}

# Test data
data_morning = [72, 68, 91, 55, 63]
data_empty   = []

print("data_morning :", analyse_sensors(data_morning))
print("data_empty   :", analyse_sensors(data_empty))

data_morning : {'average': 69.8, 'status': 'WARNING'}
data_empty   : {'average': 0, 'status': 'NO DATA'}


### Task 1.3 — Verify (5 pts)

**Output obtained:**
```
data_morning : {'average': 69.8, 'status': 'WARNING'}
data_empty   : {'average': 0,    'status': 'NO DATA'}
```

**Does status logic make sense for a factory?**  
Haan, bilkul sahi hai:
- `69.8 > 60` → **WARNING** — kuch sensors high readings de rahe hain (91 bhi hai), supervisor ko alert karna sahi hai.
- `data_empty` → **NO DATA** — ab crash nahi hota, clean message milta hai.
- CRITICAL (>85) real emergency ke liye sahi threshold hai.
- OK (<=60) normal operations ke liye theek hai.

---
## CHALLENGE 2 — NumPy Production Stats (25 pts)
### Quality Control Dashboard — Hyderabad Assembly Line

**Scenario:** Assembly line tracks product weights (grams). Quality lead needs a dashboard to flag batches outside tolerance.

### Task 2.1 — Plan (5 pts)

**`batch_stats(batch)` function:**  
Is function ko ek numpy array milti hai (product weights in grams). Hamen 4 statistics chahiye: average weight (mean), variation (std), lightest (min), heaviest (max). Numpy functions: `np.mean()`, `np.std()`, `np.min()`, `np.max()`. Return: dictionary with these 4 values.

**`quality_flag(batch, tolerance)` function:**  
Agar batch ki standard deviation (variation) tolerance (5g) se kam ya barabar ho to `'PASS'`, warna `'FAIL'`. Sirf ek numpy function chahiye: `np.std()`.

**Numpy functions needed:** `np.mean`, `np.std`, `np.min`, `np.max`, `np.array`

### Task 2.2 — Prompt & Implement (12 pts)

**Our AI Prompt:**
> *"Complete these two Python stub functions using numpy only. Do not change the function signatures or the TOLERANCE variable. Function 1: `batch_stats(batch)` should return a dictionary with keys mean, std, min, max — all as floats rounded to 2 decimal places. Function 2: `quality_flag(batch, tolerance=TOLERANCE)` should return the string 'PASS' if `np.std(batch)` is less than or equal to tolerance, otherwise return 'FAIL'."*

**Completed Functions (AI Output):**

In [2]:
# CHALLENGE 2 — Quality Control Dashboard

import numpy as np

batch_A = np.array([498, 502, 497, 503, 500, 499, 501, 505, 496, 504])
batch_B = np.array([510, 523, 488, 531, 476, 519, 482, 527, 491, 515])
batch_C = np.array([500, 500, 500, 500, 500, 500, 500, 500, 500, 500])

TOLERANCE = 5

def batch_stats(batch):
    return {
        'mean': round(float(np.mean(batch)), 2),
        'std' : round(float(np.std(batch)),  2),
        'min' : int(np.min(batch)),
        'max' : int(np.max(batch)),
    }

def quality_flag(batch, tolerance=TOLERANCE):
    return 'PASS' if np.std(batch) <= tolerance else 'FAIL'

# Run dashboard
print(f"{'Batch':<8} {'Mean':>7} {'Std':>7} {'Min':>6} {'Max':>6}  Flag")
print("-" * 48)
for name, b in [('A', batch_A), ('B', batch_B), ('C', batch_C)]:
    s    = batch_stats(b)
    flag = quality_flag(b)
    print(f"Batch {name}  {s['mean']:>7} {s['std']:>7} {s['min']:>6} {s['max']:>6}  {flag}")

Batch       Mean     Std    Min    Max  Flag
------------------------------------------------
Batch A    500.5    2.87    496    505  PASS
Batch B    506.2   19.09    476    531  FAIL
Batch C    500.0     0.0    500    500  PASS


### Task 2.3 — Interpret Results (8 pts)

| Batch | Std | Result | Reason |
|-------|-----|--------|--------|
| A | 2.87 | **PASS** | Variation 5g se kam — acceptable |
| B | 19.09 | **FAIL** | Variation bahut zyada (476g to 531g) — reject |
| C | 0.0 | **PASS** | Saari values exactly 500g |

**Why does Batch C have std = 0?**  
Batch C mein saari 10 values exactly `500` hain — koi variation nahi — isliye std = 0.

**Is that realistic in a real factory?**  
**Nahi, bilkul unrealistic hai.** Real factory machines mein hamesha thodi variation hoti hai (temperature, material, wear). std = 0 ka matlab hai ya sensor broken/stuck hai ya data manually set kiya gaya hai — dono cases investigate karne chahiye.

---
## CHALLENGE 3 — Grade System Fix (20 pts)
### University Results Processing — MUET Exam Office

**Scenario:** Borderline-pass students (exactly 50) are being marked F. Find the bug and extend with grade descriptions.

### Task 3.1 — Spot the Bug (4 pts)

**Bug location:** `elif score > 50:`  

**Explanation:**  
Jab `score = 50` hota hai, Python check karta hai `50 > 50` — ye `False` hai kyunki 50, 50 se bada nahi hai, barabar hai. Isliye ye `elif` skip ho jata hai aur `else: return 'F'` chal jata hai. 50 number wala student jo pass hai use F mil raha tha — ye bug hai.

**Correct operator:** `>=` (greater than OR equal to)  
`50 >= 50` → `True` → ab D (Pass) milega — sahi result.

### Task 3.2 — Prompt & Fix (10 pts)

**Our AI Prompt:**
> *"Fix this Python grade function — it has a bug where score=50 gets grade F because `elif score > 50` uses strict greater-than, but 50 is a passing mark. Fix it to use `>= 50`. Also: (a) add a dictionary called GRADE_DESCRIPTIONS mapping each grade letter to a description: A+: Distinction, A: Excellent, B: Very Good, C: Good, D: Pass, F: Fail. (b) Update assign_grade() to return both the letter and its description. (c) Print both for each test score."*

**AI's Improved Code:**

In [3]:
# CHALLENGE 3 — Grade System (Fixed + Extended)

GRADE_DESCRIPTIONS = {
    'A+': 'Distinction',
    'A' : 'Excellent',
    'B' : 'Very Good',
    'C' : 'Good',
    'D' : 'Pass',
    'F' : 'Fail',
}

def assign_grade(score):
    if not isinstance(score, (int, float)):
        raise ValueError(f"Score must be a number, got: {type(score).__name__}")
    if score < 0 or score > 100:
        raise ValueError(f"Score must be between 0 and 100, got: {score}")

    if score >= 90:
        letter = 'A+'
    elif score >= 80:
        letter = 'A'
    elif score >= 70:
        letter = 'B'
    elif score >= 60:
        letter = 'C'
    elif score >= 50:       # FIX: >= instead of >
        letter = 'D'
    else:
        letter = 'F'

    return letter, GRADE_DESCRIPTIONS[letter]


# Task 3.2 — Normal test scores
test_scores = [95, 83, 71, 62, 50, 49, 0]
print(f"{'Score':>6}  {'Grade':<4}  Description")
print("-" * 32)
for s in test_scores:
    letter, desc = assign_grade(s)
    print(f"{s:>6}  {letter:<4}  {desc}")

 Score  Grade  Description
--------------------------------
    95  A+    Distinction
    83  A     Excellent
    71  B     Very Good
    62  C     Good
    50  D     Pass
    49  F     Fail
     0  F     Fail


### Task 3.3 — Edge Cases (6 pts)

**Our AI Prompt for Validation:**
> *"Add input validation to the assign_grade() function. It should raise a ValueError with a clear message if the score is not between 0 and 100, or if score is not a number."*

In [4]:
# Task 3.3 — Edge Case Testing

print("Edge Case Tests:")
print("-" * 45)
for edge in [50, 100, -5, 150]:
    try:
        letter, desc = assign_grade(edge)
        print(f"score={edge:>4} -> {letter} ({desc})")
    except ValueError as e:
        print(f"score={edge:>4} -> ERROR: {e}")

Edge Case Tests:
---------------------------------------------
score=  50 -> D (Pass)
score= 100 -> A+ (Distinction)
score=  -5 -> ERROR: Score must be between 0 and 100, got: -5
score= 150 -> ERROR: Score must be between 0 and 100, got: 150


**Edge Case Analysis:**

| Score | Result | Explanation |
|-------|--------|-------------|
| 50 | D — Pass | BUG fix kaam kar raha hai |
| 100 | A+ — Distinction | Perfect score sahi handle hua |
| -5 | ValueError | Negative score impossible — validation kaam ki |
| 150 | ValueError | 100 se upar nahi ja sakta — validation kaam ki |

**Should function handle scores outside 0–100?**  
Haan, zaroor handle karna chahiye. Exam marks 0–100 se bahar nahi ho sakte — agar koi system se galat data aaye to ValueError raise karke clear message dena chahiye taake bug jaldi mil sake.

---
## CHALLENGE 4 — AI Ethics Audit (20 pts)
### Biased Hiring Algorithm — NEXUS HR System

**Scenario:** A journalist has accused NEXUS HR of university discrimination. Identify the bias, explain the harm, and rewrite it fairly.

### Task 4.1 — Run & Observe (5 pts)

In [5]:
# CHALLENGE 4 — Original Biased Algorithm

UNIVERSITY_WEIGHTS = {
    'IBA':   1.5,
    'LUMS':  1.4,
    'NUST':  1.3,
    'MUET':  0.9,
    'QUEST': 0.8,
    'other': 0.7,
}

def score_candidate_biased(name, university, gpa, years_exp):
    base   = (gpa / 4.0) * 50 + min(years_exp, 10) * 5
    weight = UNIVERSITY_WEIGHTS.get(university, UNIVERSITY_WEIGHTS['other'])
    final  = base * weight
    return {'name': name, 'score': round(final, 1), 'shortlisted': final >= 60}

candidates = [
    ('Ayesha', 'MUET',  3.8, 3),
    ('Bilal',  'LUMS',  3.2, 1),
    ('Fatima', 'QUEST', 3.9, 5),
    ('Hassan', 'IBA',   3.0, 2),
]

print("BIASED Algorithm Results:")
print("-" * 48)
print(f"{'Name':<10} {'Uni':<8} {'GPA':>5} {'Score':>7}  Shortlisted?")
print("-" * 48)
for c in candidates:
    r = score_candidate_biased(*c)
    print(f"{c[0]:<10} {c[1]:<8} {c[2]:>5} {r['score']:>7}  {r['shortlisted']}")

BIASED Algorithm Results:
------------------------------------------------
Name       Uni        GPA   Score  Shortlisted?
------------------------------------------------
Ayesha     MUET       3.8    56.2  False
Bilal      LUMS       3.2    63.0  True
Fatima     QUEST      3.9    59.0  False
Hassan     IBA        3.0    71.2  True


**Task 4.1 Observation:**

| Name | University | GPA | Score | Shortlisted |
|------|-----------|-----|-------|-------------|
| Ayesha | MUET | 3.8 | 56.2 | **NO** |
| Bilal | LUMS | 3.2 | 63.0 | YES |
| Fatima | QUEST | 3.9 | 59.0 | **NO** |
| Hassan | IBA | 3.0 | 71.2 | YES |

**Is this fair?**  
**Bilkul nahi.** Ayesha ka GPA (3.8) Hassan (3.0) se kaafi zyada hai, aur experience bhi zyada hai (3 vs 2 years) — lekin sirf MUET university ki wajah se Ayesha reject ho gayi aur Hassan (IBA, lower GPA) shortlist ho gaya. Ye system merit par nahi, university name par decisions le raha hai.

### Task 4.2 — Identify the Bias (7 pts)

**Bias Type: Proxy Discrimination (Institutional Bias)**

University name ek *"proxy"* hai — ye directly kisi characteristic ko nahi dekhta, lekin university ke zariye socio-economic background aur class ko discriminate karta hai.

**Who is harmed most?**  
MUET aur QUEST ke students — ye public sector universities hain jo Sindh ke **lower aur middle income families** ke students attend karte hain. Financially kamzor background ke talented students jo public universities mein padhte hain unhe sirf is wajah se penalize kiya ja raha hai.

**Real-world group:**  
MUET/QUEST Sindh ke rural aur semi-urban areas ke students represent karte hain jo elite private universities afford nahi kar sakte. Ye algorithm indirectly **class discrimination** aur **income inequality** ko enforce karta hai.

**Real-world harm:** Agar yeh HR system real jobs mein use ho raha ho to meritorious students job opportunities se bar ho jayenge — sirf university ke naam ki wajah se, apni actual ability ki wajah se nahi.

### Task 4.3 — Rewrite Fairly (8 pts)

**Our AI Prompt:**
> *"Rewrite the score_candidate() function to completely remove university weighting. The new fair version should only use: GPA (out of 4.0) contributing max 50 points, years_exp (capped at 10) contributing max 30 points, and an optional skills_score (0–10, default=5) from a blind interview contributing max 20 points. Shortlist threshold stays at 60. Keep the same candidates list and print new scores."*

**Fair Code + New Scores:**

In [6]:
# CHALLENGE 4 — Fair Algorithm (No University Bias)

def score_candidate_fair(name, university, gpa, years_exp, skills_score=5):
    gpa_pts   = (gpa / 4.0) * 50           # max 50 pts  — merit based
    exp_pts   = min(years_exp, 10) * 3     # max 30 pts  — experience based
    skill_pts = (skills_score / 10) * 20   # max 20 pts  — blind interview based
    final     = gpa_pts + exp_pts + skill_pts
    return {
        'name':        name,
        'university':  university,
        'score':       round(final, 1),
        'shortlisted': final >= 60
    }

print("COMPARISON: Biased vs Fair Algorithm")
print("-" * 58)
print(f"{'Name':<10} {'Uni':<8} {'GPA':>5} {'Biased Score':>13} {'Fair Score':>11}  Fair Decision")
print("-" * 58)
for c in candidates:
    biased = score_candidate_biased(*c)
    fair   = score_candidate_fair(*c)
    decision = "YES" if fair['shortlisted'] else "NO"
    print(f"{c[0]:<10} {c[1]:<8} {c[2]:>5} {biased['score']:>13} {fair['score']:>11}  {decision}")

COMPARISON: Biased vs Fair Algorithm
----------------------------------------------------------
Name       Uni        GPA  Biased Score  Fair Score  Fair Decision
----------------------------------------------------------
Ayesha     MUET       3.8          56.2        66.5  YES
Bilal      LUMS       3.2          63.0        53.0  NO
Fatima     QUEST      3.9          59.0        73.8  YES
Hassan     IBA        3.0          71.2        53.5  NO


**Fair Algorithm Results:**

| Name | GPA | Biased Score | Fair Score | Shortlisted (Fair) |
|------|-----|-------------|------------|--------------------|
| Ayesha | 3.8 | 56.2 | **66.5** | **YES** |
| Bilal | 3.2 | 63.0 | 53.0 | NO |
| Fatima | 3.9 | 59.0 | **73.8** | **YES** |
| Hassan | 3.0 | 71.2 | 53.5 | NO |

Ab merit ke hisaab se decisions ho rahe hain — highest GPA waale (Fatima 3.9, Ayesha 3.8) shortlist ho gaye.

---
## CHALLENGE 5 — Grand Build: Smart Dashboard (30 pts)
### NEXUS Smart Factory Dashboard — Build from Scratch

### Task 5.1 — Plan First (5 pts)

**Function Diagram (Pseudocode):**

```
print_dashboard(sensors, weights)
       |
       |--- analyse_sensors(sensors)
       |         -> dict {average, status}
       |         -> Uses: sum, len, if/elif/else
       |
       |--- batch_stats(weights)
       |         -> dict {mean, std, min, max}
       |         -> Uses: np.mean, np.std, np.min, np.max
       |
       |--- quality_flag(weights)
                 -> 'PASS' or 'FAIL'
                 -> Uses: np.std
```

**Output structure:**
```
===========================================
   NEXUS SMART FACTORY DASHBOARD
===========================================
[SENSOR MONITOR]
  Average, Status
  Each reading with HIGH ALERT flag

[QUALITY CONTROL]
  Mean, Std, Min, Max, Quality Flag

[MISSION SUMMARY]
  Overall status
===========================================
```

**REQ-6 checklist:**
- Function: `analyse_sensors`, `batch_stats`, `quality_flag`, `print_dashboard`
- Loop: sensor readings iterate karna (individual readings print)
- Dictionary: `stats` dict from `batch_stats`, `sensor_data` dict from `analyse_sensors`

### Task 5.2 — Prompt Chain (15 pts)

**Prompt 1 — Sensor function:**
> *"Write a Python function `analyse_sensors(readings)` that: returns `{'average': 0, 'status': 'NO DATA'}` if list is empty; calculates average; returns CRITICAL if avg>85, WARNING if avg>60, OK otherwise; returns dict with average rounded to 2 decimal places and status."*

**Prompt 2 — Batch stats + quality flag:**
> *"Write two numpy functions: (1) `batch_stats(batch)` returns a dict with mean, std, min, max all rounded to 2dp using numpy. (2) `quality_flag(batch, tolerance=5)` returns 'PASS' if np.std(batch) <= tolerance else 'FAIL'. Do not change signatures."*

**Prompt 3 — Dashboard formatter:**
> *"Write a `print_dashboard(sensors, weights)` function that calls analyse_sensors and shows average + status; loops through each sensor reading and marks ones >85 as HIGH ALERT; calls batch_stats and quality_flag on weights (as numpy array); prints a clean formatted report with section headers; ends with a MISSION SUMMARY showing overall status."*

**Final Working Code (combined from all 3 prompts):**

In [7]:
# CHALLENGE 5 — NEXUS Smart Factory Dashboard

import numpy as np

TOLERANCE = 5  # grams

# Function 1 — Sensor Monitor (from C1 fix)
def analyse_sensors(readings):
    if not readings:
        return {'average': 0, 'status': 'NO DATA'}
    average = sum(readings) / len(readings)
    if average > 85:
        status = 'CRITICAL'
    elif average > 60:
        status = 'WARNING'
    else:
        status = 'OK'
    return {'average': round(average, 2), 'status': status}

# Function 2 — Batch Stats (from C2)
def batch_stats(weights):
    arr = np.array(weights)
    return {
        'mean': round(float(np.mean(arr)), 2),
        'std' : round(float(np.std(arr)),  2),
        'min' : int(np.min(arr)),
        'max' : int(np.max(arr)),
    }

# Function 3 — Quality Flag (from C2)
def quality_flag(weights, tolerance=TOLERANCE):
    return 'PASS' if np.std(weights) <= tolerance else 'FAIL'

# Function 4 — Dashboard Printer (C5 new build)
def print_dashboard(sensors, weights):
    LINE = "=" * 58
    DASH = "-" * 44

    sensor_data = analyse_sensors(sensors)     # dict — REQ-6
    avg    = sensor_data['average']
    status = sensor_data['status']

    wt_array = np.array(weights)
    stats    = batch_stats(wt_array)           # dict — REQ-6
    flag     = quality_flag(wt_array)

    status_display = {
        'CRITICAL': '!! CRITICAL !!  (EMERGENCY - STOP LINE)',
        'WARNING' : '>> WARNING <<   (Alert supervisor)',
        'OK'      : '   OK           (All normal)',
        'NO DATA' : '   NO DATA      (No readings received)',
    }.get(status, status)

    print("\n" + LINE)
    print("       NEXUS SMART FACTORY DASHBOARD")
    print("       NEXUS Engineering Solutions")
    print(LINE)

    # Sensor Section
    print("\n  [SENSOR MONITOR - Karachi Factory Floor]")
    print("  " + DASH)
    print(f"  Total Sensors  : {len(sensors)}")
    print(f"  Average        : {avg}")
    print(f"  Status         : {status_display}")
    print(f"\n  {'No.':<5} {'Reading':>8}  Alert")
    print("  " + "-" * 28)
    for i, val in enumerate(sensors, 1):       # loop — REQ-6
        alert = "<-- HIGH ALERT" if val > 85 else ""
        print(f"  {i:<5} {val:>8}  {alert}")

    # Quality Control Section
    print(f"\n  [QUALITY CONTROL - Hyderabad Assembly Line]")
    print("  " + DASH)
    print(f"  Mean Weight    : {stats['mean']} g")
    print(f"  Std Deviation  : {stats['std']} g")
    print(f"  Min Weight     : {stats['min']} g")
    print(f"  Max Weight     : {stats['max']} g")
    print(f"  Tolerance      : +/-{TOLERANCE} g")
    print(f"  Quality Flag   : {flag}")

    # Summary Section
    if status == 'CRITICAL' or flag == 'FAIL':
        overall = "IMMEDIATE ACTION REQUIRED"
    elif status == 'WARNING':
        overall = "MONITOR CLOSELY - WARNING ACTIVE"
    else:
        overall = "ALL SYSTEMS NOMINAL"

    print("\n" + LINE)
    print("  MISSION SUMMARY")
    print(LINE)
    print(f"  Sensor Status  : {status}")
    print(f"  Quality Flag   : {flag}")
    print(f"  Overall Status : {overall}")
    print(LINE)


# Task 5.3 — Test with given data
sensors = [72, 68, 91, 55, 63, 88, 74, 66, 95, 58]
weights = [498, 502, 497, 503, 500, 499, 501, 505, 496, 504]

print_dashboard(sensors, weights)


       NEXUS SMART FACTORY DASHBOARD
       NEXUS Engineering Solutions

  [SENSOR MONITOR - Karachi Factory Floor]
  --------------------------------------------
  Total Sensors  : 10
  Average        : 73.0
  Status         : >> WARNING <<   (Alert supervisor)

  No.    Reading  Alert
  ----------------------------
  1           72  
  2           68  
  3           91  <-- HIGH ALERT
  4           55  
  5           63  
  6           88  <-- HIGH ALERT
  7           74  
  8           66  
  9           95  <-- HIGH ALERT
  10          58  

  [QUALITY CONTROL - Hyderabad Assembly Line]
  --------------------------------------------
  Mean Weight    : 500.5 g
  Std Deviation  : 2.87 g
  Min Weight     : 496 g
  Max Weight     : 505 g
  Tolerance      : +/-5 g
  Quality Flag   : PASS

  MISSION SUMMARY
  Sensor Status  : WARNING
  Quality Flag   : PASS
  Overall Status : MONITOR CLOSELY - WARNING ACTIVE


### Task 5.3 — Test & Reflect (10 pts)

**Dashboard Output Analysis:**

| Metric | Value | Meaning |
|--------|-------|---------|
| Sensor Average | 73.0 | WARNING — 60 se upar |
| High Sensors | 3 (91, 88, 95) | 85 se upar — HIGH ALERT |
| Batch Std | 2.87 | PASS — tolerance ke andar |
| Overall | MONITOR CLOSELY | Sensor warning active hai |

**Reflection — What surprised us most about vibecoding:**  
Sabse surprising ye tha ke agar prompt clear aur specific ho, to AI bilkul sahi structure de deta hai — lekin hamen har output carefully verify karna parta hai. AI kabhi kabhi logic galat kar deta hai (jaise boundary conditions) jo test karne se hi pata chalta hai. Vibecoding sirf code likhna nahi — prompt likhna aur output verify karna bhi skill hai.

**What we would change with more time:**
1. CSV file se real sensor data load karna
2. Timestamp add karna — kis waqt reading aayi
3. Email/SMS alert system jab status CRITICAL ho
4. Historical trend graph banana (matplotlib)

---
## MISSION DEBRIEF — Team Reflection

**1. Which challenge was hardest? What made it difficult?**  
Challenge 4 (Ethics) sabse mushkil tha — code likhna asaan tha, lekin bias ko technically explain karna aur real-world harm samajhna zyada sochne wali cheez thi.

**2. Did AI ever give you wrong code? How did you catch it?**  
Haan — pehle attempt mein C1 ka fix mein AI ne `if len(readings) == 0` likha jo kaam karta hai lekin Pythonic nahi hai. Humne `if not readings:` se replace kiya. Output run karke verify kiya.

**3. What was your best prompt of the session? Why did it work well?**  
C4 ka fair rewrite prompt sabse acha tha — usme clearly scoring breakdown (50+30+20) batai gayi thi. Specific numbers dene se AI ne exact formula implement kiya bina guess kiye.

**4. Where did you see AI bias or automation bias in today's challenges?**  
C4 mein clearly institutional bias thi — university name ka proxy use karna. Automation bias risk: agar koi HR team is algorithm ka output blindly maane bina check kiye, to talented MUET students hamesha reject hote rahenge.

**5. One thing we will do differently next time:**  
Pehle test cases likho, phir code — Test-Driven Development (TDD) approach. Is tarah AI ka output automatically verify hoga aur bugs jaldi milenge.

---
**NEXUS Engineering Solutions — AI Vibecoding Mission**  
*Conscious AI Use: Prompt. Verify. Understand.*